File: hermes/src/producer.rs
use rdkafka::config::ClientConfig;
use rdkafka::producer::{FutureProducer, FutureRecord, Producer as RdProducer};
use rdkafka::util::Timeout;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::Duration;
use tracing::{info, warn};

pub trait EventProducer: Send + Sync + 'static {
    fn produce(&self, topic: &str, key: String, payload: Vec<u8>);
    fn is_ready(&self) -> bool;
}

pub struct KafkaProducer {
    inner: FutureProducer,
    ready: Arc<AtomicBool>,
}

impl KafkaProducer {
    pub fn new(brokers: &str) -> Result<Self, rdkafka::error::KafkaError> {
        // Config values ported verbatim from botserver/server/producer.js
        let producer: FutureProducer = ClientConfig::new()
            .set("bootstrap.servers", brokers)
            .set("retry.backoff.ms", "200")
            .set("message.send.max.retries", "10")
            .set("socket.keepalive.enable", "true")
            .set("queue.buffering.max.messages", "100000")
            .set("queue.buffering.max.ms", "1000")
            .set("batch.num.messages", "1000000")
            // Explicit murmur2_random to match librdkafka/Java default and node-rdkafka behavior
            .set("partitioner", "murmur2_random")
            .create()?;
        Ok(KafkaProducer {
            inner: producer,
            ready: Arc::new(AtomicBool::new(false)),
        })
    }

    /// Fetches broker metadata as a readiness probe; flips the ready flag on success.
    pub fn wait_ready(&self, timeout_ms: u64) -> Result<(), rdkafka::error::KafkaError> {
        self.inner
            .client()
            .fetch_metadata(None, Duration::from_millis(timeout_ms))?;
        self.ready.store(true, Ordering::Relaxed);
        info!("producer ready");
        Ok(())
    }
}

impl EventProducer for KafkaProducer {
    fn produce(&self, topic: &str, key: String, payload: Vec<u8>) {
        let producer = self.inner.clone();
        let topic = topic.to_string();
        // Create record inside the spawned task so all references are to owned data
        // with no lifetime escaping the method. Timeout::After(Duration::ZERO) is
        // non-blocking: if the queue is full it fails immediately rather than retrying.
        tokio::spawn(async move {
            let record = FutureRecord::to(&topic).key(&key).payload(&payload);
            if let Err((e, _)) = producer.send(record, Timeout::After(Duration::ZERO)).await {
                warn!(key = %key, "Kafka delivery failed: {}", e);
            }
        });
    }

    fn is_ready(&self) -> bool {
        self.ready.load(Ordering::Relaxed)
    }
}

File: replybot/lib/producer.js
const Kafka = require('node-rdkafka')

const producer = new Kafka.Producer({
  'metadata.broker.list': process.env.KAFKA_BROKERS,
  'retry.backoff.ms': 200,
  'message.send.max.retries': 10,
  'request.required.acks': 1,
  'socket.keepalive.enable': true,
  'queue.buffering.max.messages': 100000,
  // linger.ms — was 1000, ~1s of the measured 1669ms delivery floor. 10 rather
  // than librdkafka's 5ms default is just headroom: a turn's commands are
  // produced microseconds apart (publishCommands) and batch under any value.
  // VIR-36 tracks the retry-ordering guarantee that batching currently masks.
  'queue.buffering.max.ms': 10,
  'batch.num.messages': 1000000
}, {}, {});

producer.connect()
producer.setPollInterval(1000)

producer.on('event.error', err => {
  console.error('Error from producer');
  throw err;
})

const producerReady = new Promise((resolve, reject) => {
  const timeout = setTimeout(() => {
    reject(new Error('Unable to connect to kafka producer'))
  }, process.env.KAFKA_CONNECTION_TIMEOUT || 30000)


  producer.on('ready', () => {
    console.log('producer ready')
    clearTimeout(timeout)
    resolve()
  })

}).catch(err => {
  setTimeout(() => { throw err })
})

module.exports = { producer, producerReady }
